pkg/pillar/types/global.go

```diff
@@ -103,312 +106,410 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
-		newgc.ResetIfCloudGoneTime = GlobalConfigMinimums.ResetIfCloudGoneTime
+	specMap.GlobalSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddTriStateItem(key GlobalSettingKey, defaultTriState TriState) {
+	configItem := ConfigItemSpec{
+		Key:             string(key),
+		ItemType:        ConfigItemTypeTristate,
+		TriStateDefault: defaultTriState,
 	}
-	if newgc.FallbackIfCloudGoneTime < GlobalConfigMinimums.FallbackIfCloudGoneTime {
-		log.Warnf("Enforce minimum FallbackIfCloudGoneTime received %d; using %d",
-			newgc.FallbackIfCloudGoneTime, GlobalConfigMinimums.FallbackIfCloudGoneTime)
-		newgc.FallbackIfCloudGoneTime = GlobalConfigMinimums.FallbackIfCloudGoneTime
+	specMap.GlobalSettings[key] = configItem
+	log.Debugf("Added tristate item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddAgentSettingStringItem(key AgentSettingKey, defaultString string) {
+	configItem := ConfigItemSpec{
+		ItemType:      ConfigItemTypeString,
+		Key:           string(key),
+		StringDefault: defaultString,
 	}
-	if newgc.MintimeUpdateSuccess < GlobalConfigMinimums.MintimeUpdateSuccess {
-		log.Warnf("Enforce minimum MintimeUpdateSuccess received %d; using %d",
-			newgc.MintimeUpdateSuccess, GlobalConfigMinimums.MintimeUpdateSuccess)
-		newgc.MintimeUpdateSuccess = GlobalConfigMinimums.MintimeUpdateSuccess
+	specMap.AgentSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, item config.ConfigItem) error {
+	// legacy per-agent setting key debug.<agentname>.xxx
+	// new per-agent setting key agent.<agentname>.debug.xxx
+	var configSpec ConfigItemSpec
+	agentSetting := false
+	globalSetting := false
+	key := item.Key
+	components := strings.Split(key, ".")
+	if _, ok := specMap.AgentSettings[AgentSettingKey(key)];
+	strings.HasPrefix(key, "agent") || strings.HasPrefix(key, "debug") && ok {
+		if strings.HasPrefix(key, "agent") && len(components) > 2 {
+			components = components[2:]
+		} else if strings.HasPrefix(key, "debug") && len(components) > 2 {
+			components[1] = ""
+		} else {
+			return fmt.Errorf("Unable to find agent name for per-agent setting. Key: %s", key)
+		}
+		key = strings.Join(components, ".")
+		agentSetting = true
+	} else if _, ok := specMap.GlobalSettings[GlobalSettingKey(key)]; ok {
+		globalSetting = true
+	} else {
+		return fmt.Errorf("Item is neither a global nor a per-agent setting. Key: %s", key)
+	}
+	if agentSetting {
+		configSpec = specMap.AgentSettings[AgentSettingKey(key)]
+		agentName := components[1]
+		val, err := parseConfigItemToValue(configSpec, item)
+		if err == nil {
+			configMap.AgentSettings[agentName][AgentSettingKey(key)] = val
+		} else {
+			return err
+		}
+	} else if globalSetting {
+		configSpec = specMap.GlobalSettings[GlobalSettingKey(key)]
+		val, err := parseConfigItemToValue(configSpec, item)
```

Comment:
val, err := configSpec.ParseValue(valueStr)
   -- See the comment for parseConfigItemToValue function.. it is better to be a method of configSpec - as it is using the spec and parsing he given string and returning ConfigItemValue..